internal/bminventory/inventory.go

```
@@ -1809,3 +1811,66 @@ func (b *bareMetalInventory) validateDNSRecords(clusterName string, domain *dnsD
 	return validations.CheckDNSRecordsExistence(vipAddresses, domain.ID, domain.Provider)
 }
+
+func ipAsUint(ipStr string, log logrus.FieldLogger) uint64 {
+	parts := strings.Split(ipStr, ".")
+	if len(parts) != 4 {
+		log.Warnf("Invalid ip %s", ipStr)
+		return 0
+	}
+	var result uint64 = 0
+	for _, p := range parts {
+		result = result << 8
+		converted, err := strconv.ParseUint(p, 10, 64)
+		if err != nil {
+			log.WithError(err).Warnf("Conversion of %s to uint", p)
+			return 0
+		}
+		result += converted
+	}
+	return result
+}
+
+func applyLimit(ret models.FreeAddressesList, limitParam *int64) models.FreeAddressesList {
+	if limitParam != nil && *limitParam >= 0 && *limitParam < int64(len(ret)) {
+		return ret[:*limitParam]
+	}
+	return ret
+}
+
+func (b *bareMetalInventory) getFreeAddresses(params installer.GetFreeAddressesParams, log logrus.FieldLogger) middleware.Responder {
+	var hosts []*models.Host
+	err := b.db.Select("free_addresses").Find(&hosts, "cluster_id = ? and status in (?)", params.ClusterID.String(), []string{host.HostStatusInsufficient, host.HostStatusKnown}).Error
+	if err != nil {
+		return common.NewApiError(http.StatusInternalServerError, errors.Wrapf(err, "Error retreiving hosts for cluster %s", params.ClusterID.String()))
+	}
+	if len(hosts) == 0 {
+		return common.NewApiError(http.StatusNotFound, errors.Errorf("No hosts where found for cluster %s", params.ClusterID))
+	}
+	resultingSet := network.MakeFreeAddressesSet(hosts, params.Network, params.Prefix, log)
+
+	ret := models.FreeAddressesList{}
+	for a := range resultingSet {
+		ret = append(ret, a)
+	}
+
+	// Sort addresses
+	sort.Slice(ret, func(i, j int) bool {
+		return ipAsUint(ret[i].String(), log) < ipAsUint(ret[j].String(), log)
+	})
+
+	ret = applyLimit(ret, params.Limit)
+
+	return installer.NewGetFreeAddressesOK().WithPayload(ret)
+}
+
+func (b *bareMetalInventory) GetFreeAddresses(ctx context.Context, params installer.GetFreeAddressesParams) middleware.Responder {
```

I think all those functions should return an error not `middleware.Responder`
The main functions should take the error and if it's type is `common.ApiErrorResponse` then return it else generate default error.